Simplenote/src/main/java/com/automattic/simplenote/AddTagActivity.java

```diff
@@ -95,56 +77,87 @@ public void onClick(View v) {
-                        finishAfterTransition();
-                    } catch (BucketObjectNameInvalid bucketObjectNameInvalid) {
-                        showDialogError();
-                    }
+                    TextInputEditText tagInput = binding.tagInput;
+                    String tag = tagInput.getText() != null ? tagInput.getText().toString() : "";
+                    viewModel.saveTag(tag);
                 }
             }
         );
-        mButtonPositive.setEnabled(false);
+        binding.buttonPositive.setEnabled(false);
+    }
+
+    private void setupLayout() {
+        int widthScreen = getResources().getDisplayMetrics().widthPixels;
+        int widthMargin = (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 56, getResources().getDisplayMetrics());
+        int widthMaximum = (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 312, getResources().getDisplayMetrics());
+        LinearLayout layout = binding.layout;
+        layout.getLayoutParams().width = Math.min(widthMaximum, widthScreen - widthMargin);
+        layout.requestLayout();
+
+        ((View) layout.getParent()).setOnClickListener(
+            new View.OnClickListener() {
+                @Override
+                public void onClick(View view) {
+                    finishAfterTransition();
+                }
+            }
+        );
+        layout.setOnClickListener(null);
+
+        MorphSetup.setSharedElementTransitions(this, layout, getResources().getDimensionPixelSize(R.dimen.corner_radius_dialog));
+    }
+
+    private void setObservers() {
+        // Setup observer to show an error in case the tag name is not valid
+        viewModel.getTagError().observe(this, new Observer<Integer>() {
+            @Override
+            public void onChanged(Integer error) {
+                if (error == null) {
+                    binding.tagLayout.setError(null);
+                    binding.buttonPositive.setEnabled(true);
+                } else {
+                    String errorMessage = getString(error);
+                    binding.tagLayout.setError(errorMessage);
+                    binding.buttonPositive.setEnabled(false);
+                }
+            }
+        });
+
+        // Setup observer to show or hide the keyboard
+        viewModel.getShowKeyboard().observe(this, new Observer<Boolean>() {
+            @Override
+            public void onChanged(Boolean showKeyboard) {
+                if (showKeyboard) {
+                    binding.tagInput.requestFocus();
+                    DisplayUtils.showKeyboard(binding.tagInput);
+                } else {
+                    DisplayUtils.hideKeyboard(binding.tagInput);
+                }
+            }
+        });
+
+        // Setup observer for result of saving a tag
+        viewModel.isResultOK().observe(this, new Observer<Boolean>() {
+            @Override
+            public void onChanged(Boolean tagSaved) {
+                if (tagSaved) {
```

Comment:
I think this moves the logic unnecessarily to the activity. I'd rather either use 2 live data instances for finish and dialog actions, or even better have something like `val event: LiveData<Event>` where event is an enum with values like `FINISH` and `SHOW_ERROR`. 
Moreover this solution can crash because the live data event could be consumed more than once (it will most likely not happen but it's possible). I'd create a Github issue for the future to introduce something like a generic event that could be consumed only once (check WPAndroid for the approach).

Reply:
I have added an `event` field that handle the following events:

- CLOSE: closes the activity without saving a tag
- FINISH: closes the activity after the tag was saved correctly and send the result of the activity as OK
- SHOW_ERROR: shows a dialog with a error message in case the tag could not be saved

Fixed in ff0380dcc28559b1f4652e58177aa23988a8e8e0.